### PR TITLE
3.17 Pilot - ATF Refactor

### LINKED ISSUE #6772: 3.17 - Refactor Activation process
# Description

This issue focuses on the proposed refactor of the `Activation` process within the LCP feature. The goal is to make the `Activation` process more generic and reusable for other features. This involves moving the `Activation` process from the `AboveTheFold` namespace to a new namespace called `PerformanceHints` in the `Common`  directory. The `PerformanceHints` namespace will serve as a central location for processes that are generic and shared across features.

# Scope a solution:

- Create a new namespace `PerformanceHints` in the `Common` namespace if it doesn't exist already.
- Move the Activation folder from `AboveTheFold` to `PerformanceHints`.
- Remove all reference of the Admin, WarmUp, Cron, and Activation processes from AboveTheFold\ServiceProvider
- Update the Common\PerformanceHints\ServiceProvider if created and if not create it and reflect the new location of the Activation processes.
- Thoroughly test the `Activation` process in its new location to ensure it works as expected and maintains its functionality.
- Update any documentation or comments in the codebase to reflect the change in location of the Activation process.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### LINKED ISSUE #6773: 3.17 - Refactor Admin
# Description:

This issue focuses on the proposed refactor of the Admin process within the `AboveTheFold` feature. The goal is to make the `Admin` process more generic and reusable for other features. This involves moving the `Admin` process from the `AboveTheFold` namespace to the `PerformanceHints` namespace within `Common`. The `Common` namespace will serve as a central location for processes that are generic and shared across features.

# Scope a solution:

- Move the `Admin` folder from `AboveTheFold` to `PerformanceHints` within `Common`.
- Remove all references to the Admin module in `AboveTheFold\ServiceProvider`
- Update the `Common\PerformanceHints\ServiceProvider` if created and if not create it and reflect the new location of the Admin module.
- Thoroughly test the `Admin` process in its new location to ensure it works as expected and maintains its functionality.
- Update any documentation or comments in the codebase to reflect the change in location of the `Admin` process.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### LINKED ISSUE #6774: 3.17 - Refactor WarmUp
# Description:

This issue focuses on the proposed refactor of the `WarmUp` process within the `AboveTheFold` feature. The goal is to make the `WarmUp` process more generic and reusable for other features. This involves moving the `WarmUp` process from the `AboveTheFold` namespace to the `PerformanceHints` namespace within `Common`. The `Common` namespace will serve as a central location for processes that are generic and shared across features.

# Scope a solution:

- Move the `WarmUp` [[folder](https://github.com/wp-media/wp-rocket/blob/77ed4f2a0d94d4ec729b3fa133902758b7772e7f/inc/Engine/Media/AboveTheFold/WarmUp)](https://github.com/wp-media/wp-rocket/blob/77ed4f2a0d94d4ec729b3fa133902758b7772e7f/inc/Engine/Media/AboveTheFold/WarmUp) from `AboveTheFold` to `PerformanceHints` within `Common`.
- Remove all references to the WarmUp process in `AboveTheFold\ServiceProvider`
- Update the `Common\PerformanceHints\ServiceProvider` if created and if not create it and reflect the new location of the WarmUp processes.
- Update all references to the `WarmUp` process in the codebase to reflect its new location in the `PerformanceHints` namespace within `Common`.
- Thoroughly test the `WarmUp` process in its new location to ensure it works as expected and maintains its functionality.
- Update any documentation or comments in the codebase to reflect the change in location of the `WarmUp` process.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### LINKED ISSUE #6775: 3.17 - Refactor ServiceProvider
# Description:

This issue focuses on the proposed refactor of the `ServiceProvider` within the `AboveTheFold` feature. The goal is to update the `ServiceProvider` to reflect the changes made in the `Admin`, `WarmUp`, `Cron`, and `Activation` processes. These processes have been moved to the `PerformanceHints` namespace within `Common`. The `Common` namespace serves as a central location for processes that are generic and shared across features.

# Scope a solution:

- Remove all reference of the `Admin`, `WarmUp`, `Cron`, and `Activation` processes from `AboveTheFold\ServiceProvider`
- Update the `Common\PerformanceHints\ServiceProvider` if created and if not create it and reflect the new location of the `Admin`, `WarmUp`, `Cron`, and `Activation` processes.
- Thoroughly test the `ServiceProvider` to ensure it works as expected and maintains its functionality.
- Update any documentation or comments in the codebase to reflect the changes made in the `ServiceProvider`.

# Acceptance Criteria
- All tests should be green and passing.

### LINKED ISSUE #6776: 3.17 - Refactor AJAX Process
# Description:

This issue focuses on the proposed refactor of the `AJAX` process within the `AboveTheFold` feature. The goal is to make the `AJAX` process more generic and reusable for other features. This involves moving the `AJAX` process from the `AboveTheFold` namespace to the `PerformanceHints` namespace. The `PerformanceHints` namespace will serve as a central location for processes that are generic and shared across features. Additionally, an interface for the AJAX controller will be created, which will be implemented in `PerformanceHints/AboveTheFold/AJAX/Controller.php`.

# Scope a solution:

image

- Create an interface `ControllerInterface` in `Common\PerformanceHints\AJAX` namespace. This interface will define the methods that any AJAX controller should implement. Define the methods below in the interface:
  - `add_data()`
  - `check_data()`

```php
<?php
declare(strict_types=1);

namespace WP_Rocket\Engine\Common\PerformanceHints\AJAX;

interface ControllerInterface {
    public function add_data();
    public function check_data();
}
```

- Implement the interface in `AboveTheFold/AJAX/Controller` based on [current Controller](https://github.com/wp-media/wp-rocket/blob/3db051802769e624cfd1ab97bf5a540a0fc9e542/inc/Engine/Media/AboveTheFold/AJAX/Controller.php)

```php
<?php
declare(strict_types=1);

namespace WP_Rocket\Engine\AboveTheFold\AJAX;

use WP_Rocket\Engine\PerformanceHints\ControllerInterface;
use WP_Rocket\Engine\Media\AboveTheFold\Database\Queries\AboveTheFold as ATFQuery;
use WP_Rocket\Engine\Common\Context\ContextInterface;
use WP_Rocket\Engine\Optimization\UrlTrait;

class Controller implements ControllerInterface {
    public function add_data() {
        // Implementation of the method...
    }

    public function check_data() {
        // Implementation of the method...
    }
}
```

- In the `PerformanceHints/AJAX/Subscriber`, update the code to reflect the newly renamed functions
- Here, we would be passing the factories as an array via DI

```php
<?php
declare(strict_types=1);

namespace WP_Rocket\Engine\Common\PerformanceHints\AJAX;

use WP_Rocket\Event_Management\Subscriber_Interface;

private $factories = [];

class Subscriber implements Subscriber_Interface {
    public function __construct( array $factories ) {
        $this->factories = $factories;
    }

    public static function get_subscribed_events(): array {
		return [
			// The rest of the events goes here.
		];
	}

    public function add_data() {
        foreach( $this->factories as $factory ) {
            $factory->controller->add_data();
        }
    }

    public function check_data() {
        foreach( $this->factories as $factory ) {
            $factory->controller->check_data();
        }
    }
}
```
- Remove `AboveTheFold\AJAX\Subscriber` as we already have a unified one in `Common\PerformanceHints\AJAX` now
- Thoroughly test the `AJAX` process in its new location to ensure it works as expected and maintains its functionality.
- Update any documentation or comments in the codebase to reflect the change in location of the `AJAX` process and the renaming of the methods.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### LINKED ISSUE #6777: 3.17 Sub tasks - Factories Creation
# Description
This is a sub-task for the planned refactor of `AboveTheFold` in preparation for the `LRC` feature. Here we will be creating an abstract factory that can be used by any of the feature concrete factories so this way they can be used conveniently.

# Scope a solution:
image

- Add a new namespace - `Common\PerformanceHints`
    - Create a new interface called `AbstractFactory` having methods:
        - `ajax` to return `AjaxInterface` (Interface will be created when adding the ajax modules)
        - `frontend` to return `FrontendInterface` (Interface will be created when adding the frontend modules)
        - `table` to return `TableInterface` (Interface will be created when adding the table modules)
        - `query` to return `QueryInterface` (Interface will be created when adding the query modules)
- Create a Concrete factory in `AboveTheFold` namespace, you can call it Factory which will implement creation methods of the AbstractFactory Above
- This Factory will hold the objects readily for each methods above passed via `DI`.
- Lastly, update `AboveTheFold\ServiceProvider` with the new factory and make it shared.

### LINKED ISSUE #6778: 3.17 Sub tasks - Frontend Refactoring
# Description
This is a sub-task for the planned refactor of the `AboveTheFold` in preparation for the `LRC` feature. Here we will be refactoring this to make use of an abstract and concrete product to make the application not focus on the feature that needs to be processed but rather the process.

# Scope a solution:
image

- Add a new namespace - `Common\PerformanceHints\Frontend`
    - Create an interface in the above namespace, this will have generic method names that will be implemented.
    - Create a subscriber that will bind events to concrete controller referenced from the factories objects passed via `DI`.
- We need to refactor this controller in `WP_Rocket\Engine\Media\AboveTheFold\Frontend` to implement the generic methods from the respective interface we created above.
- Move the `Subscriber` contents to the new `Subscriber` in `Common\PerformanceHints\Frontend` and then delete it.
- Update `AboveTheFold\ServiceProvider`.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### LINKED ISSUE #6779: 3.17 Sub tasks - Database Refactoring
# Description
This is a sub-task for the planned refactor of the AboveTheFold in preparation for the LRC feature. Here we will be changing the Database modules a bit, we’ll be focusing on `AboveTheFold\Database\Tables` & `AboveTheFold\Database\Queries`

# Scope a solution:
image

- Add a new namespace - `Common\PerformanceHints\Database`
    - Create 2 interfaces:
        - Table
            - We already have a similar interface in `Engine\Common\Database` but there is a class specific method in `AboveTheFold\Database\Tables` i.e `truncate_atf_table`, this kind of method cannot be defined in the current interface as it will trigger an error in the rest of classes that implements that interface and also we don’t use all defined methods in that interface for the AboveTheFold feature. So this new table interface should only contain generic methods like truncate_table, remove_all_completed_rows
        - Query - Define generic methods here
- In `AboveTheFold\Database\Tables`
    - Update the Class to implement the `Table` Interface methods created above. 
- In `AboveTheFold\Database\Queries`
    - Update the Class to implement the `Query` Interface methods created above. 
- Update the Concrete factory in AboveTheFold\Factory to return their respective objects for table and queries.

### Additional Context
Something else to consider would be to have abstract table & queries class for even better extensibility that way when we starting working on LRC, we will not have to define methods again for pretty much the same DB operations, but again this should be accessed during this refactor.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### LINKED ISSUE #6781: 3.17 - Refactor Cron
# Description:

This issue focuses on the proposed refactor of the Admin process within the `AboveTheFold` feature. The goal is to make the `Cron` process more generic and reusable for other features. This involves moving the `Cron` process from the `AboveTheFold` namespace to the `PerformanceHints` namespace within `Common`. The `Common` namespace will serve as a central location for processes that are generic and shared across features.

# Scope a solution:

- Move the `Cron` folder from `AboveTheFold` to `PerformanceHints` within `Common`.
- Remove all references to the Cron module in `AboveTheFold\ServiceProvider`
- Update the `Common\PerformanceHints\ServiceProvider` if created and if not create it and reflect the new location of the Cron module.
- Thoroughly test the `Cron` process in its new location to ensure it works as expected and maintains its functionality.
- Update any documentation or comments in the codebase to reflect the change in location of the `Cron` process.

# Acceptance Criteria
- There should be no regression for this module
- All tests should be green and passing.

### PR DESCRIPTION
**Pull request by jeawhanlee**
# Description

Closes https://github.com/wp-media/wp-rocket/issues/6777 https://github.com/wp-media/wp-rocket/issues/6775 https://github.com/wp-media/wp-rocket/issues/6781 https://github.com/wp-media/wp-rocket/issues/6778 https://github.com/wp-media/wp-rocket/issues/6776 https://github.com/wp-media/wp-rocket/issues/6764 https://github.com/wp-media/wp-rocket/issues/6772 https://github.com/wp-media/wp-rocket/issues/6774 https://github.com/wp-media/wp-rocket/issues/6779 https://github.com/wp-media/wp-rocket/issues/6773

### Technical documentation

Refactor OCI to share Performance hints modules also updated hooks to be generic and deprecated old ones
```
rocket_lcp_width_threshold -> rocket_performance_hints_optimization_width_threshold
rocket_lcp_height_threshold -> rocket_performance_hints_optimization_height_threshold
rocket_lcp_delay -> rocket_performance_hints_optimization_delay
rocket_after_clear_atf -> rocket_after_clear_performance_hints_data
rocket_atf_warmup_links_number ->  rocket_performance_hints_warmup_links_number,
```

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies
None

## Risks
None

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Documentation

- [ ] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [ ] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [ ] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [ ] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.
- [ ] I listed the introduced external dependencies explicitely on the PR.
- [ ] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [x] I handled errors when needed.
- [ ] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [ ] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [ ] I explicitely mentioned performance risks in the PR.
- [ ] I explicitely mentioned security risks in the PR.

### CODE REVIEW

**Comment by jeawhanlee** [comment]
Blocking till [this](https://github.com/wp-media/wp-rocket/pull/6830) is ready.

File: package.json [hunk 1]
```diff
@@ -49,7 +49,7 @@
     "watchify": "^4.0.0",
     "webpack": "^5.76.0",
     "webpack-cli": "^4.9.1",
-	"wp-rocket-scripts": "^1.0.1",
+	"wp-rocket-scripts": "github:wp-media/rocket-scripts#enhancement/beacon-script-refactoring-phase-2",
```
Comment (Miraeld) [inline package.json]: Will it still be this branch or it will be a dedicated version ?

**Review by Miraeld — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@wp-media/engineering-plugin-team Thanks for the PR. Can't delete 3.17 zip, it will result in a fatal error. can you please check :pray: 
[debug.log](https://github.com/user-attachments/files/16440349/debug.log)
steps:
1- fresh install and activation of the zip
2- deactivate WPR
3- delete WPR

**Comment by MathieuLamiot** [comment]
Can someone check this before moving forward?
https://wp-media.slack.com/archives/CUT7FLHF1/p1722546721389749?thread_ts=1722433782.954279&cid=CUT7FLHF1

**Review by MathieuLamiot — CHANGES_REQUESTED** [review]
Blocked by: https://github.com/wp-media/wp-rocket/issues/6838
See https://wp-media.slack.com/archives/CUT7FLHF1/p1722596982917919?thread_ts=1722433782.954279&cid=CUT7FLHF1

**Review by MathieuLamiot — APPROVED** [review]

**Review by Miraeld — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
Not blocked anymore.

**Review by Mai-Saad — APPROVED** [review]
Failed tests are either accepted for now or have GH issue.
- Accepted issues as per @piotrbak 
1- On local env, Fetch the home page (added to logs but not DB) when fresh install or trigger clear critical images. Thanks @hanna-meda 
2- Fetched Urls from home  (added to log)  are double when updated from version < 3.16  (DB has no duplication)

[testrail-report-626.pdf](https://github.com/user-attachments/files/16526821/testrail-report-626.pdf)

**Pull request merged**